simba/trip.py

```diff
@@ -1,54 +1,77 @@
-from datetime import datetime, timedelta
+from datetime import datetime
 
 
 class Trip:
-    def __init__(self, rotation, departure_time, departure_name,
-                 arrival_time, arrival_name, distance, **kwargs):
+    def __init__(self, departure_time, departure_name,
+                 arrival_time, arrival_name, distance,
+                 line=None, temperature=None, level_of_loading=None, mean_speed=None,
+                 height_diff=None, station_data=None,
+                 charging_type=None, vehicle_type=None,
+                 ):
+        self.departure_time = departure_time
         self.departure_name = departure_name
-        self.departure_time = datetime.fromisoformat(departure_time)
-        self.arrival_time = datetime.fromisoformat(arrival_time)
+        self.arrival_time = arrival_time
         self.arrival_name = arrival_name
         self.distance = float(distance)
-        self.line = kwargs.get('line', None)
-        self.temperature = kwargs.get('temperature', None)
+
+        # optional / computed
+        self.height_diff = height_diff  # m
+        self.level_of_loading = level_of_loading  # [0-1]
+        self.line = line
+        self.mean_speed = mean_speed  # km/h
+        self.temperature = temperature  # deg C
+
+        # needed to update rotation properties when adding trip
+        self.charging_type = charging_type  # oppb/depb/None
+        self.vehicle_type = vehicle_type
+
+        # set when added to rotation
+        # needed for consumption calculation and other references, like schedule
+        # !Circular reference! problematic for garbage collector in older Python versions
+        self.rotation = None
+
+        # set after calculating consumption
+        self.consumption = None  # kWh
+        self.delta_soc = None
+
+        # ---- cast / compute options ---- #
+        # cast times to datetime
+        if type(departure_time) is str:
+            self.departure_time = datetime.fromisoformat(departure_time)
+        if type(arrival_time) is str:
+            self.arrival_time = datetime.fromisoformat(arrival_time)
+
```

Comment:
Imo this gives the Trip constructor to much responsibility. type checking and casting should be done outside of the constructor. Helper/ Builder functions can take care of the appropriate casting. In this case maybe the add_trips_from_dict function. same applies for the following lines.